### PR TITLE
Use DESTDIR when checking systemv script exists

### PR DESCRIPTION
**Pull request by kingosticks**
There looks to be a missing `$DESTDIR` in the systemv init script part of the makefile. 

I noticed this when trying to [create a Debian package](https://github.com/kingosticks/shairport-sync/tree/debian) and had tried to use a `dh_installinit --onlyscripts` rule. 

The call to `update-rc.d` then also gave problems.  Would it be reasonable to only perform that command when $(DESTDIR) is unset? Something like:
```
[ -n $(DESTDIR) ] || update-rc.d shairport-sync defaults 90 10
```

### CODE REVIEW

**Comment by mikebrady** [comment]
Hi Nick. Thanks for all your work on this – it would be just great to have a Debian package for Shairport Sync. I had a look at it about a year ago and discovered at the time that using OpenSSL made it ineligible, so hence the move to PolarSSL. Then, I never went back to it, so thanks again for your efforts.

I hope you'll pardon my ignorance when I ask a couple of questions about the pull request:
* I understand that the `install-exec-hook` stuff is only executed if you do a `make install`. So my first question is whether it is appropriate to be using the `make install` stuff inside a Debian package – should it not be using its own installation scripts?
* If the answer to the above is that it is indeed "legitimate" to use the `install-exec-hook` stuff, my second question is if your check for $(DESTDIR) being unset is a sort of a hack?

Anyway, apologies again for asking these questions – I just want to ensure we don't over-complicate things for ourselves going forward.

Regards
Mike

**Comment by p3ck** [comment]
Is this update-rc.d a new thing?  I've never seen it in other packages. 

Usually make install just installs packages and should be implemented in a way that works in a "chroot" (ie: DESTDIR).

**Comment by kingosticks** [comment]
Hi Mike,
Yes I saw your [comments](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=743638) regarding the licensing, to be honest I didn't quite understand since shairport-sync itself is not GPL but it doesn't matter since using polarSSL is fine. 

The bad news is that this is the first attempt at Debian packaging (which I am doing as part of https://github.com/woutervanwijk/Pi-MusicBox/issues/6) so all I know is from what I have read in the last few months.

`dh_auto_install` invokes `make install` and I believe this is the sort of thing is designed to help with. The [init script section](https://www.debian.org/doc/manuals/maint-guide/dother.en.html#initd) states (my emphasis)
> If the upstream init script looks fine and **installs in the right place** you still need to set up the rc* symlinks.

Alternatively, I could build with `--without-configfiles` and add an entry in [debian/install](https://www.debian.org/doc/manuals/maint-guide/dother.en.html#install) to do the copying.  I suppose either is fine but I'm not sure which is less complicated.

Is it a hack? Yeh I think so, but as @p3ck says, `make install` should really just work in a chroot and currently (by default) it doesn't.

**Comment by kingosticks** [comment]
Sorry, I'll rephrase, maybe it does actually work but `update-rc.d` might just fail silently but it probably shouldn't be invoked and go around (trying) to changing the host system.

**Comment by mikebrady** [comment]
`update-rc.d` is a utility for ensuring startup scripts are installed and updated correctly – see, for example, http://www.debuntu.org/how-to-managing-services-with-update-rc-d/.

TBH, I think it would be more prudent to go the `--without-configfiles` and the [default] `--without-systemv` route, as your script would then be independent of any changes made to those in the future.

**Comment by mikebrady** [comment]
@p3ck, could you clarify what you mean by "Usually make install just installs packages and should be implemented in a way that works in a "chroot" (ie: DESTDIR)." please?

**Comment by kingosticks** [comment]
@mikebrady yep ok sounds good to me. I can open a separate issue regarding any other Debian package issues I come across. Hopefully the original fix is still worth merging.

**Comment by p3ck** [comment]
Hi Mike,

I just mean that normally make install can be redirected to another
$DESTDIR.  And any commands executed during that would honor it.

I'm still not completely sure why the command needs to be run.  is it
similar to the following with systemd which turns the service on for the
next reboot?

systemctl enable shairport-sync

If so then I wouldn't expect a make install to enable a service for me..  I
would expect it to simply install it. :-)

But I don't feel too strongly about it..  it only affects the systemv
scripts right?

Thanks!

**Comment by mikebrady** [comment]
@kingosticks and @p3ck, thanks for your explanations and your patience, and thanks for the patch. In regard to `update-rc.d`, I don't think it's quite like `systemd`'s `systemctl enable`. I believe that what it does is add and remove links and scripts so that the the app may be launched and terminated properly at boot, shutdown etc. AFAIK, it doesn't actually directly influence the run state of the app, so it is in the spirit of a regular installer. It does only affect System V scripts.

**Comment by rkam** [comment]
Hey guys,

FYI:

**update-rc.d** is the equivalent of **systemctl**.

It's convoluted, the history of all this, but basically:

These are two (of several) ways of managing services under linux: *System V init-based* style and *Systemd* style.

### SystemV

The most obvious indicator of *SystemV* is that the config scripts live in */etc/init.d*

As Mike points out, the services are configured by creating links between */etc/init.d* and */etc/rcN.d* 

When the system changes state, it looks at the links in the appropriate */etc/rcN.d* directories to figure out what to do.

- Now, **update-rc.d** is **a** tool to manage those links.

(and **service** and **chkconfig** are alternatives to **update-rc.d** - *within* the *SystemV* subsystem.)

### Systemd

*Systemd* is an alternative to the entire *SystemV init-based* subsystem.

Since it replaces the entire subsystem, it does things very differently.  Most obvious is that it does not use */etc/init.d*

- **systemctl** is the utility to use to configure the services.

---
Hence **systemctl** is the equivalent to **update-rc.d** (and the pair: **service** and **chkconfig**).

Finally, *systemd* is the "new way to go" and most linux distributions have moved to it (https://en.wikipedia.org/wiki/Systemd#Adoption_and_reception).

--
Here's a terse explanation of the different subsystems: https://en.wikipedia.org/wiki/Linux*startup*process#Init*process
(Note that it describes the original *SystemV init-based* subsystem (early UNIX) - which used */etc/rcN* FILES and not */etc/rcN.d* directories. - The way they both work is the same, only the location of the configuration changed - essentially, configuration for a particular service moved out of the single rc*N* file and into it's own file in the rc*N*.d directory.).

I can't find a mapping of **update-rc.d** to **systemctl**, but this might be useful - it's between **service/chkconfig** and **systemctl** : https://fedoraproject.org/wiki/SysVinit*to*Systemd*Cheatsheet (among others)
Basically, **service** manages the service and **chkconfig** manages the configuration of the service and **update-rc.d** does both (as does **systemctl**).

**Comment by rkam** [comment]
Btw, looking at the makefile, it doesn't look to me that the service configuration is done for the systemd case.

(there should be a *systemctl enable shairport-sync* to be equivalent to the SystemV case)

**Comment by p3ck** [comment]
Hello everyone,

I disagree that make install should also enable the service.  After all the
make target is called install, not install+enable. :-)

Thanks,
Bill.

**Comment by mikebrady** [comment]
Hi everyone. Thanks for this interesting discussion. I am still learning about it, and I would really like to understand it better so that, for example, an installer would create (or check the existence of) a low-privileged user and group at the time on installation. The startup script would then launch Shairport Sync to run under the account of that user, maybe do a `chroot`, etc.

Just to go back to `update-rc.d`, to the best of my knowledge it only installs stuff, or else checks and modifies a prior installation. Again, AFAIK, it does not start the service or restart an existing one. We may be at cross-purposes about the term "enable": `update-rc.d` does "enable" the service, but all that means is that it installs the links so that the service will be launched on reboot; it does not start or restart the service. To do that, you'd use the `service` utility.

Am I right about that? Does that address the issue you raise, @p3ck?

**Comment by p3ck** [comment]
Hi Mike,

I think it might help to think about distro packaging.  In order to package
the program we need to install the binaries, shared libs and whatever other
files are needed for the program to run into a temporary top level dir (ie:
/usr, /etc, etc..).  Everything installed in there is going to be put in
the package.  Now lets look at the systemV init script, we need the file in
/etc/init.d but we don't want to package the symlinks in /etc/rc5.d for
example.  Why?  Because we shouldn't be packaging files that are not
managed by the package manager. If we did then if I disabled the service
and then did a rpm --verify on the package it would complain that a file is
missing.

Does that make sense?

To enable the service or to add a user is usually done in %post scripts of
the package.  (I think deb has the same thing?)

I understand you want to make it as easy for users to use the program but
usually that is why distributions package them into rpms or debs. :-)

HTH,
Bill

**Comment by rkam** [comment]
You are both correct in all of your points. :)  I think the confusion is from the transition in state your project has made. - Mike, your project is so successful that others want to use it in theirs.  (Congratulations!)

So, while you have (correctly) been working from the point of view of people who *git* it, build it and install it, @p3ck is in another realm - packaging, which has a different set of rules and conventions.

In other words, this project is in the middle ground between being a small project that people always get the source for and build, and people that don't ever get source - they just use apt-get, etc.

So, up until now, it was nice to have the links created so the service starts and stops correctly when the system changes state (e.g. reboots).  It's desirable and perfectly fine.

But, as Bill  says, it interferes with package management, which is what he wants to do.

Mike, you are also correct in that **update-rc.d** does not actually start the service.  It creates the links which enable the service to start (or stop) when the system changes state (e.g. boots or shuts down).

Nevertheless, that part is really part of _system configuration_. - You are telling the system how it should function, rather than installing all it needs to accomplish that.

So, to be package friendly, it shouldn't do the **update-rc.d** command as part of _make install_.  As Bill says, it should only install the parts, not configure them.
(btw, I wasn't advocating keeping the service configure step in the install - I was just pointing out that it the two SystemX parts weren't doing the same steps - i.e. systemd won't start the service on reboot).

For the easiest, short-term change, I'd suggest making yet another option to make that configures the service to run.  Move the **update-rc.d** to it and tell (git) users to use that new option and package users wouldn't.  (I'll see about making a patch..)

(sorry for being so long winded)
---
    @mikebrady wrote: an installer would create (or check the existence of) a low-privileged user and group at the time on installation. The startup script would then launch Shairport Sync to run under the account of that user, ...

Yes, this is fine and a common thing to do.  (Up to where I cut it off)

    @mikebrady  wrote: maybe do a chroot, etc.

This confuses me.  Why would you do that?

(If you are thinking of @p3ck's comments about chroot with respect to DESTDIR, that's something else.)

**Comment by mikebrady** [comment]
Hi everyone – sorry for the delay in replying and thanks for all your inputs. It seems that the "cleanest" thing to do is to remove `update-rc.d` from the `make install` script completely and add it to the [human] instructions for installation. The thing is that the existing instructions for completing installation need more work, both for `System V` and `systemd` systems, e.g. for adding low-privileged users.

Would there be general approval for that approach?

BTW that idea of chrooting the running program is something I think I've seen in a few places, maybe (?) for example, Avahi. I understand makes it a little harder to break into the rest of the system if the program is compromised.

**Comment by kingosticks** [comment]
I think that makes the most sense.

**Comment by rkam** [comment]
Yes, that's the correct way.

Additional make targets that do the configuration part is fine too - Completely up to you.
